### DISCUSSION TITLE
ADFS MFA with Smartcards (Yubikey)

### DISCUSSION THREAD

**Discussion by derSchweiger** [neos-sdi/adfsmfa#347] [Q&A] [answered]
Hi,

I am currently evaluating Yubikeys to see if it is possible to seamlessly integrate them into a hybrid (more on-premises, but still hybrid) Active Directory environment. Therefore, I have deployed user certificates from our on-premises PKI to these Yubikeys and the logon to the local workstations works just fine. 
The next goal is to use these user certificates in combination with the Yubikeys (smartcards) to log in to ADFS. Unfortunately, this doesn't work very well - at least I couldn't get it to run at this point. My approach:
- Use the standard Active Directory login as primary factor
- Utilise your ADFS MFA plugin as second factor and trigger the biometrics MFA provider (instead of TOTP)

When trying to use the biometrics MFA provider, I can't tell Windows not to use Windows Hello for Business. As soon as I try to register the biometric device (in this case the Yubikey) it always triggers WHfB.
Is there a way to use your MFA plugin with Yubikeys without enrolling WHfB? Especially in relation to non-domain joined Windows PCs?

**Comment by redhook62** [comment]
Hi

Biometrics is based on the "open standard": WebAuthN/FIDO2, WHfB is specific to Microsoft. If you use Yubico products, it is not the smart card or user certificate part of your AD that must be used, but the FIDO part of these dongles (Yubico Security Key 4, Yubikey 4 and 5, Yubikey Bio, etc.)
If you have configured the MFA plug-in in ADDS mode, be careful in the case of joint use with WHfN not to use the Attributes model for Windows 2016, because in this case we use a multi-value AD attribute that is used primarily by WHfB.
Of course WebAuthN biometrics does not need WHfB, moreover this protocol advantageously replaces WHfB for Web authentication

**Comment by derSchweiger** [comment]
Hi @redhook62,

thank you as always for your quick response!
> Biometrics is based on the "open standard": WebAuthN/FIDO2, WHfB is specific to Microsoft. If you use Yubico products, it is not the smart card or user certificate part of your AD that must be used, but the FIDO part of these dongles (Yubico Security Key 4, Yubikey 4 and 5, Yubikey Bio, etc.)

Understood and totally makes sense! Mixed that up.

> If you have configured the MFA plug-in in ADDS mode, be careful in the case of joint use with WHfN not to use the Attributes model for Windows 2016, because in this case we use a multi-value AD attribute that is used primarily by WHfB.

I guess you are talking about the "msDS-KeyCredentialLink" attribute? Yes, this is only used for WHfB and ADFS MFA is configured for a different AD attribute.


> Of course WebAuthN biometrics does not need WHfB, moreover this protocol advantageously replaces WHfB for Web authentication

If that's the way to get around WHfB, then that might be the right choice for me. As far as I've understood the WebAuthN configuration for your MFA plugin correctly, I need to set "_None_" or at least "_Cross-Platform_" as "_Authenticator Attachment_". But as soon as I set this, I receive the following error message while adding a new biometric device: _Error details: Unable to read beyond the end of the stream._ Same error message in the eventviewer. 
Current config:
```powershell
PS C:\Users\Administrator.WINDOOF> Get-MFAProvider -ProviderType Biometrics
DirectLogin                 : False
Timeout                     : 60000
TimestampDriftTolerance     : 300000
ChallengeSize               : 16
ServerDomain                : censored.dev
ServerName                  : auth
ServerIcon                  :
Origin                      : https://auth.censored.dev
PinRequirements             : Null
UseNickNames                : True
ForbiddenBrowsers           : ie;samsung;nokia
InitiatedBrowsers           : safari
NoCounterBrowsers           : safari
Enabled                     : True
EnrollWizard                : True
EnrollWizardDisabled        : False
LockUserOnDefaultProvider   : False
ForceWizard                 : Disabled
PinRequired                 : False
IsRequired                  : True
FullQualifiedImplementation :
Parameters                  :

PS C:\Users\Administrator.WINDOOF> Get-MFASecurity -Kind BIOMETRIC
AuthenticatorAttachment         : Empty
AttestationConveyancePreference : Direct
UserVerificationRequirement     : Preferred
Extensions                      : True
UserVerificationMethod          : True
RequireResidentKey              : False
ConstrainedMetadataRepository   : False

``` 

Any idea why this occurs?

**Comment by redhook62** [comment]
For a Yubikey, you need cross-platform because this key is external, put something other than "Direct"

**Comment by derSchweiger** [comment]
Yep, already tried that but receive the same error message:
```powershell
PS C:\Users\Administrator.WINDOOF> Get-MFASecurity -Kind BIOMETRIC
AuthenticatorAttachment         : CrossPlatform
AttestationConveyancePreference : None
UserVerificationRequirement     : Preferred
Extensions                      : True
UserVerificationMethod          : True
RequireResidentKey              : False
ConstrainedMetadataRepository   : False
``` 

Also tried "Indirect" instead of "None" for "AttestationConveyancePreference".

As soon as I click on "Continue" while on "Enroll your biometric device":
![image](https://github.com/neos-sdi/adfsmfa/assets/79589007/a33c32b8-d3e3-4abe-9446-8c297af1aa39)
Tried it with different browsers, as well as internal/external and different clients.

**Reply by derSchweiger** [reply]
Nevermind, seems to be related to my test environment. Have tested the same setup on another deployment and there it's just working fine!

**Comment by redhook62** [comment, marked as answer]
Hi,

Indeed, this seems linked to your testing platform.

Below, a WebAuthN configuration that works perfectly with Windows Hello, Yubikeys and different dongles. and without WHfB configured.

![image](https://github.com/neos-sdi/adfsmfa/assets/26826945/baee9597-5f77-4353-b6e2-163961395870)
